python/ql/src/Security/CWE-327/Ssl.qll

```diff
@@ -0,0 +1,186 @@
+import python
+import semmle.python.ApiGraphs
+import semmle.python.dataflow.new.internal.Attributes as Attributes
+import TlsLibraryModel
+
+class SSLContextCreation extends ContextCreation {
+  override CallNode node;
+
+  SSLContextCreation() { this = API::moduleImport("ssl").getMember("SSLContext").getACall() }
+
+  override DataFlow::CfgNode getProtocol() {
+    result.getNode() in [node.getArg(0), node.getArgByName("protocol")]
+  }
+}
+
+class SSLDefaultContextCreation extends ContextCreation {
+  SSLDefaultContextCreation() {
+    this = API::moduleImport("ssl").getMember("create_default_context").getACall()
+  }
+
+  // Allowed insecure versions are "TLSv1" and "TLSv1_1"
+  // see https://docs.python.org/3/library/ssl.html#context-creation
+  override DataFlow::CfgNode getProtocol() { none() }
+}
+
+class WrapSocketCall extends ConnectionCreation {
+  override CallNode node;
+
+  WrapSocketCall() { node.getFunction().(AttrNode).getName() = "wrap_socket" }
```

Comment:
I'm not too fond of this implementation, which is just look at an `AttrNode`. Could we some solution where we know what API-graph nodes represent instances of SSLContext, and then do `sslContextInstance().getMember("wrap_socket").getACall()`?

Reply:
I used a type tracker, I think there might not be an API node for the instances?